Remove checks for outdated GCC versions

PBfordev opened the pull request:
wxWidgets now supports only GCC 4 and newer, so there is no reason to keep code specific for older GCC versions.

BTW, looking at the compiler names and versions, I wonder if [this whole block](https://github.com/PBfordev/wxWidgets/blob/a7df23d43c4f5c321e1ccd5b59ca260c547e3d03/tests/archive/archivetest.cpp#L27) could not be significantly shortened. MSVC 2003 is the oldest supported version so `_MSC_VER >= 1310` can be removed, but I have no idea what that `__WIN64__`  check has to do with member templates nor do I know whether those HP and Sun compilers are still being supported...

PBfordev committed:
Remove checks for outdated GCC versions

wxWidgets now supports only GCC 4 and newer, so there is no reason
to keep code specific for older GCC versions.

PBfordev committed:
Remove checks for wxNEEDS_CHARPP

wxNEEDS_CHARPP was needed only for now unsupported GCC versions.

vadz commented on include/wx/cpp.h:
```diff
@@ -140,20 +140,6 @@
 
 #ifdef HAVE_VARIADIC_MACROS
```
Note to self: we probably should get rid of this one as well. We still need to have checks for old MSVC versions, but there is almost certainly no need for configure test.

vadz commented on include/wx/msw/gccpriv.h:
```diff
@@ -77,9 +77,7 @@
-    #if ( __GNUC__ > 2 )
-        #define HAVE_W32API_H
-    #endif
+    #define HAVE_W32API_H
```
Probably could get rid of this one entirely and just always include `<w32api.h>` here.

PBfordev replied:
I know nothing about how Cygwin or WINE, so I'd rather not touch that part.

vadz replied:
I'm going to remove it because not having to test for it in configure and define it in makefiles is a nice gain, so I'll just push the extra commits in this PR in a moment.

vadz reviewed (approved):
Thanks!

We should definitely apply this, but we probably could simplify things even a bit more here.

vadz committed:
Remove HAVE_W32API_H definition and checks

All still supported versions of MinGW/Cygwin provide w32api.h file, so
there is no need to test for it.

This allows to simplify the code, but also remove the definition of
HAVE_W32API_H from bake- and makefiles and this, in turn, allows to get
rid of extra flags in MinGW format entirely, as we don't support gcc
2.95 for which they were originally needed neither.

vadz committed:
Remove variadic macros test from configure

This is useless, all still supported compilers except ancient MSVS
versions (for which configure is not used anyhow) support variadic
macros, so don't waste time testing for them.

Note that the checks for HAVE_VARIADIC_MACROS in the sources are still
left because it is still possible to explicitly disable variadic macros
support using --disable-vararg_macros for strict C++98 compatibility.

PBfordev commented:
I suppose testing for variadic macros support can be removed from CMake as well?

E.g. 
https://github.com/wxWidgets/wxWidgets/blob/master/build/cmake/setup.h.in#L747
https://github.com/wxWidgets/wxWidgets/blob/master/build/cmake/setup.cmake#L135

vadz commented:
> I suppose testing for variadic macros support should be removed from CMake as well?
> 
> E.g.
> https://github.com/wxWidgets/wxWidgets/blob/master/build/cmake/setup.h.in#L747
> https://github.com/wxWidgets/wxWidgets/blob/master/build/cmake/setup.cmake#L135

I didn't do it because I wasn't sure if it's not possible to use CMake with ancient MSVS versions that don't support it. Probably not, but who knows.

PBfordev commented:
> .... I wasn't sure if it's not possible to use CMake with ancient MSVS versions...
Ah, it seems variadic macros are not supported in MSVS 2003; sorry for the noise.
